### ISSUE TITLE
Game crashed creating a raw Storage Vessel

### ISSUE THREAD

**Issue by risingfish** [anegostudios/VintageStory-Issues#2284]
### Game Version

v1.17.9

### Platform

Windows

### Modded

Vanilla

### SP/MP

Multiplayer

### Description

Was creating a storage vessel and the game crashed:

```
Running on 64 bit Windows with 32 GB RAM 
Game Version: v1.17.9 (Stable)
11/10/2022 7:37:28 PM: Critical error occurred
Loaded Mods: game@1.17.9, creative@1.17.9, survival@1.17.9
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at Vintagestory.GameContent.BlockEntityClayForm.OnCopyLayer(Int32 layer) in VSSurvivalMod\BlockEntity\BEClayForm.cs:line 375
   at Vintagestory.GameContent.BlockEntityClayForm.OnUseOver(IPlayer byPlayer, Vec3i voxelPos, BlockFacing facing, Boolean mouseBreakMode) in VSSurvivalMod\BlockEntity\BEClayForm.cs:line 189
   at Vintagestory.GameContent.ItemClay.OnHeldInteractStop(Single secondsUsed, ItemSlot slot, EntityAgent byEntity, BlockSelection blockSel, EntitySelection entitySel) in VSSurvivalMod\Item\ItemClay.cs:line 133
   at Vintagestory.Client.NoObf.SystemMouseInWorldInteractions.HandleHandInteraction(Single dt) in VintagestoryLib\Client\Systems\Player\MouseInWorldInteractions.cs:line 317
   at Vintagestory.Client.NoObf.SystemMouseInWorldInteractions.OnGameTick(Single dt) in VintagestoryLib\Client\Systems\Player\MouseInWorldInteractions.cs:line 165
   at Vintagestory.Common.EventManager.TriggerGameTick(Int64 ellapsedMilliseconds, IWorldAccessor world) in VintagestoryLib\Common\EventManager.cs:line 106
   at Vintagestory.Client.NoObf.ClientMain.MainRenderLoop(Single dt) in VintagestoryLib\Client\ClientMain.cs:line 1200
   at Vintagestory.Client.NoObf.ClientMain.MainGameLoop(Single deltaTime) in VintagestoryLib\Client\ClientMain.cs:line 1104
   at Vintagestory.Client.GuiScreenRunningGame.RenderToPrimary(Single dt) in VintagestoryLib\Client\MainMenu\Screens\GuiScreenRunningGame.cs:line 125
   at Vintagestory.Client.ScreenManager.Render(Single dt) in VintagestoryLib\Client\ScreenManager.cs:line 680
   at Vintagestory.Client.ScreenManager.OnNewFrame(Single dt) in VintagestoryLib\Client\ScreenManager.cs:line 624
   at Vintagestory.Client.NoObf.ClientPlatformWindows.window_RenderFrame(Object sender, FrameEventArgs e) in VintagestoryLib\Client\ClientPlatform\GameWindow.cs:line 131
   at System.EventHandler`1.Invoke(Object sender, TEventArgs e)
   at OpenTK.GameWindow.RaiseRenderFrame(Double elapsed, Double& timestamp) in C:\Users\Nexrem\Desktop\transfer\opentk\src\OpenTK\GameWindow.cs:line 479
   at OpenTK.GameWindow.DispatchRenderFrame() in C:\Users\Nexrem\Desktop\transfer\opentk\src\OpenTK\GameWindow.cs:line 454
   at OpenTK.GameWindow.Run(Double updates_per_second, Double frames_per_second) in C:\Users\Nexrem\Desktop\transfer\opentk\src\OpenTK\GameWindow.cs:line 375
   at Vintagestory.Client.ClientProgram.Start(ClientProgramArgs args, String[] rawArgs)
   at Vintagestory.ClientNative.CrashReporter.Start(ThreadStart start) in VintagestoryLib\Client\ClientPlatform\ClientNative\CrashReporter.cs:line 93
```

### Logs

Server didn't crash so those don't have anything, but here's more client log info.

```
10.11.2022 18:54:16 [Notification] Received 2421 item types from server
10.11.2022 18:54:16 [Notification] Loaded 8547 block types from server
10.11.2022 18:54:16 [Notification] Composed 1 4096x4096 entities texture atlases from 184 textures (last textureid = 162)
10.11.2022 18:54:17 [Notification] Collected 1581 shapes to tesselate.
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/angelica.png' not found (defined in Item game:herbbundle-angelica).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/arnica.png' not found (defined in Item game:herbbundle-arnica).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/chicory.png' not found (defined in Item game:herbbundle-chicory).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/coltsfoot.png' not found (defined in Item game:herbbundle-coltsfoot).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/coriander.png' not found (defined in Item game:herbbundle-coriander).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/foxglove.png' not found (defined in Item game:herbbundle-foxglove).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/horsehound.png' not found (defined in Item game:herbbundle-horsehound).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/marigold.png' not found (defined in Item game:herbbundle-marigold).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/mugwort.png' not found (defined in Item game:herbbundle-mugwort).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/rockrose.png' not found (defined in Item game:herbbundle-rockrose).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/rosemary.png' not found (defined in Item game:herbbundle-rosemary).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/starflower.png' not found (defined in Item game:herbbundle-starflower).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/yarrow.png' not found (defined in Item game:herbbundle-yarrow).
10.11.2022 18:54:17 [Notification] Composed 1 4096x4096 items texture atlases from 1246 textures (last textureid = 163)
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/angelica.png' not found (defined in Item code game:herbbundle-angelica).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/arnica.png' not found (defined in Item code game:herbbundle-arnica).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/chicory.png' not found (defined in Item code game:herbbundle-chicory).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/coltsfoot.png' not found (defined in Item code game:herbbundle-coltsfoot).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/coriander.png' not found (defined in Item code game:herbbundle-coriander).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/foxglove.png' not found (defined in Item code game:herbbundle-foxglove).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/horsehound.png' not found (defined in Item code game:herbbundle-horsehound).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/marigold.png' not found (defined in Item code game:herbbundle-marigold).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/mugwort.png' not found (defined in Item code game:herbbundle-mugwort).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/rockrose.png' not found (defined in Item code game:herbbundle-rockrose).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/rosemary.png' not found (defined in Item code game:herbbundle-rosemary).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/starflower.png' not found (defined in Item code game:herbbundle-starflower).
10.11.2022 18:54:17 [Warning] Texture asset 'game:textures/item/resource/herb/yarrow.png' not found (defined in Item code game:herbbundle-yarrow).
10.11.2022 18:54:17 [Notification] Composed 1 4096x4096 blocks texture atlases from 3863 textures (last textureid = 164)
10.11.2022 18:54:18 [Notification] Server assets loaded
10.11.2022 18:54:18 [Notification] Loaded Shaderprogramm for render pass instanced.
10.11.2022 18:54:18 [Notification] Loading world map cache db...
10.11.2022 18:54:18 [Notification] Initialized Music Engine
10.11.2022 18:54:18 [Notification] Texture size is 32 so decal atlas size of 128x128 should suffice
10.11.2022 18:54:18 [Notification] Blocks tesselated
10.11.2022 18:54:18 [Notification] Received level finalize
10.11.2022 18:54:19 [Notification] Loaded Shaderprogramm for render pass clouds.
10.11.2022 18:54:19 [Notification] Track music/through-the-grass-survive.ogg now started
10.11.2022 18:54:31 [Notification] Finished fully loading sounds (async)
10.11.2022 18:58:00 [Error] The dialog GuiDialogBlockEntityRecipeSelector requested focus, but was not added yet. Missing call to api.Gui.RegisterDialog()
10.11.2022 18:58:51 [Error] The dialog GuiDialogBlockEntityRecipeSelector requested focus, but was not added yet. Missing call to api.Gui.RegisterDialog()
10.11.2022 19:01:10 [Notification] Track music/to-dawn.ogg now started
10.11.2022 19:08:05 [Notification] Track music/summer-day.ogg now started
10.11.2022 19:15:00 [Error] The dialog GuiDialogBlockEntityRecipeSelector requested focus, but was not added yet. Missing call to api.Gui.RegisterDialog()
10.11.2022 19:15:47 [Notification] Track music/creating.ogg now started
10.11.2022 19:28:04 [Notification] Track music/through-the-grass-survive.ogg now started
10.11.2022 19:28:51 [Error] The dialog GuiDialogBlockEntityRecipeSelector requested focus, but was not added yet. Missing call to api.Gui.RegisterDialog()
10.11.2022 19:34:18 [Notification] Track music/to-dawn.ogg now started
10.11.2022 19:34:59 [Error] The dialog GuiDialogBlockEntityRecipeSelector requested focus, but was not added yet. Missing call to api.Gui.RegisterDialog()
10.11.2022 19:37:28 [Fatal] Game Version: v1.17.9 (Stable)
11/10/2022 7:37:28 PM: Critical error occurred
Loaded Mods: game@1.17.9, creative@1.17.9, survival@1.17.9
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at Vintagestory.GameContent.BlockEntityClayForm.OnCopyLayer(Int32 layer) in VSSurvivalMod\BlockEntity\BEClayForm.cs:line 375
   at Vintagestory.GameContent.BlockEntityClayForm.OnUseOver(IPlayer byPlayer, Vec3i voxelPos, BlockFacing facing, Boolean mouseBreakMode) in VSSurvivalMod\BlockEntity\BEClayForm.cs:line 189
   at Vintagestory.GameContent.ItemClay.OnHeldInteractStop(Single secondsUsed, ItemSlot slot, EntityAgent byEntity, BlockSelection blockSel, EntitySelection entitySel) in VSSurvivalMod\Item\ItemClay.cs:line 133
   at Vintagestory.Client.NoObf.SystemMouseInWorldInteractions.HandleHandInteraction(Single dt) in VintagestoryLib\Client\Systems\Player\MouseInWorldInteractions.cs:line 317
   at Vintagestory.Client.NoObf.SystemMouseInWorldInteractions.OnGameTick(Single dt) in VintagestoryLib\Client\Systems\Player\MouseInWorldInteractions.cs:line 165
   at Vintagestory.Common.EventManager.TriggerGameTick(Int64 ellapsedMilliseconds, IWorldAccessor world) in VintagestoryLib\Common\EventManager.cs:line 106
   at Vintagestory.Client.NoObf.ClientMain.MainRenderLoop(Single dt) in VintagestoryLib\Client\ClientMain.cs:line 1200
   at Vintagestory.Client.NoObf.ClientMain.MainGameLoop(Single deltaTime) in VintagestoryLib\Client\ClientMain.cs:line 1104
   at Vintagestory.Client.GuiScreenRunningGame.RenderToPrimary(Single dt) in VintagestoryLib\Client\MainMenu\Screens\GuiScreenRunningGame.cs:line 125
   at Vintagestory.Client.ScreenManager.Render(Single dt) in VintagestoryLib\Client\ScreenManager.cs:line 680
   at Vintagestory.Client.ScreenManager.OnNewFrame(Single dt) in VintagestoryLib\Client\ScreenManager.cs:line 624
   at Vintagestory.Client.NoObf.ClientPlatformWindows.window_RenderFrame(Object sender, FrameEventArgs e) in VintagestoryLib\Client\ClientPlatform\GameWindow.cs:line 131
   at System.EventHandler`1.Invoke(Object sender, TEventArgs e)
   at OpenTK.GameWindow.RaiseRenderFrame(Double elapsed, Double& timestamp) in C:\Users\Nexrem\Desktop\transfer\opentk\src\OpenTK\GameWindow.cs:line 479
   at OpenTK.GameWindow.DispatchRenderFrame() in C:\Users\Nexrem\Desktop\transfer\opentk\src\OpenTK\GameWindow.cs:line 454
   at OpenTK.GameWindow.Run(Double updates_per_second, Double frames_per_second) in C:\Users\Nexrem\Desktop\transfer\opentk\src\OpenTK\GameWindow.cs:line 375
   at Vintagestory.Client.ClientProgram.Start(ClientProgramArgs args, String[] rawArgs)
   at Vintagestory.ClientNative.CrashReporter.Start(ThreadStart start) in VintagestoryLib\Client\ClientPlatform\ClientNative\CrashReporter.cs:line 93
Crash written to file at "C:\Users\Zac Feuerborn\AppData\Roaming\VintagestoryData\Logs\client-crash.txt"

10.11.2022 19:37:28 [Notification] Exiting game now. Server running=False. Exit reason: Game crashed
```

**Comment by Craluminum2413** [comment]
It looks like you hit the "17 layer"

**Comment by risingfish** [comment]
Do I get a prize? :)

**Comment by Craluminum2413** [comment]
Does it crash every time you creating it?

**Comment by risingfish** [comment]
No, I've creative a number of them since then and they were fine. It seems
to be a one off event.

**Comment by noelle-lavenza** [comment]
`NextNotMatchingRecipeLayer` is able to return 16 under certain conditions, even though valid indices for clayforming layers go from 0 to 15.
https://github.com/anegostudios/vssurvivalmod/blob/9f2f42da65ac74d1aff50f480c6c38769c87890a/BlockEntity/BEClayForm.cs#L296
If it's called while layer 16 (index 15) is completed, potentially due to a server tick running long and finishing the clayform (which is server-side only) being delayed, it pass 16 to `OnCopyLayer`, causing an IndexOutOfRangeException on line 384:
https://github.com/anegostudios/vssurvivalmod/blob/9f2f42da65ac74d1aff50f480c6c38769c87890a/BlockEntity/BEClayForm.cs#L384
This race condition results in the observed error for any clayforms that are a full 16 layers tall, under conditions of server lag or if you get really really unlucky. A potential solution would be to just return 15 instead of 16 by default in `NextNotMatchingRecipeLayer`, or check in OnCopyLayer if the layer passed is in bounds.

**Comment by noelle-lavenza** [comment]
Yes, actually! I ran a local dedicated server, connected to it, and created a clay vessel while using Clumsy to simulate conditions of severe server lag (500ms lag, 50% dropped packets). Finishing a clay vessel using layer duplicate, particularly when clicking very fast like I was, caused the attached crash which is identical to the one observed in the original post.
[client-crash.txt](https://github.com/anegostudios/VintageStory-Issues/files/10202578/client-crash.txt)

**Comment by radfast** [comment]
Fixed in upcoming 1.19.5-stable, apologies it took us so long to get to this

**Comment by risingfish** [comment]
No problem, I am happy I was able to help make the game a little better than it already is!
